WIP: Starting to add demo nft module (#297)
Sovereign-Labs/sovereign: examples/demo-nft-module/README.md

# How to create new module for Sovereign Rollup

# Getting Started

Module is a separate crate that needs to implement `sov_modules_api::Module` trait and it's own change logic based on input messages.

Let's start with defining basic dependencies in `Cargo.toml`:

```toml
[dependencies]
anyhow = { workspace = true }
sov-modules-api = { path = "../../module-system/sov-modules-api", default-features = false }
sov-modules-macros = { path = "../../module-system/sov-modules-macros" }
```

## Module definition

NFT module is defined as following:

```rust
use sov_modules_api::Context;
use sov_modules_macros::ModuleInfo;

#[derive(ModuleInfo, Clone)]
pub struct NonFungibleToken<C: Context> {
    #[address]
    pub address: C::Address,

    #[state]
    pub(crate) admin: sov_state::StateValue<C::Address>,

    #[state]
    pub(crate) owners: sov_state::StateMap<u64, C::Address>,
}
```

Where it has:

> 1. Address. Each module has to have address. TBD: Why, and how it is defined and used.

Reviewer: Having addresses allows modules to interact with each other (in the same way that smart contracts can interact with each other in Ethereum). For example, you can use the `bank.transfer` function to transfer tokens to a module.

IIRC, the address is calculated deterministically based on the module name. CC @bkolad

> 1. Address. Each module has to have address. TBD: Why, and how it is defined and used.

Reviewer 2: A module address is similar to a `smart contract` address in Ethereum. So, for example, a module can receive money. The address is calculated from a hash of the module prefix. This means that:
- The module address is unique 
- No one knows private key that produces this address

2. Several state attributes. In this case it is address of admin, and map of token id to owner address. Token id is u64 for simplicity.

> TBD: What is context, what does it hold?

Author: This question

Reviewer: Context is a trait which allows the runtime to pass verified data to modules during execution. Currently the only required method in Context `sender()` which returns the address of the person who initiated the transaction (the signer).

Context also inherits the `Spec` trait, which defines the concrete types used by the rollup for `Hash`ing, persistant data `Storage`, digital `Signature`s, and `Addresses`. The `Spec` trait allows rollups to easily tailor themselves to different Zkvms. By being generic over a `Spec`, a rollup can ensure that any potentially SNARK-unfriendly cryptography can be easily swapped out.

sov-modules-macros = { git = "https://github.com/Sovereign-Labs/sovereign.git", branch = "main" }
```

## Module definition

NFT module is defined as following:

```rust
use sov_modules_api::Context;
use sov_modules_macros::ModuleInfo;

#[derive(ModuleInfo, Clone)]
pub struct NonFungibleToken<C: Context> {
    #[address]
    pub address: C::Address,

    #[state]
    pub(crate) admin: sov_state::StateValue<C::Address>,

    #[state]
    pub(crate) owners: sov_state::StateMap<u64, C::Address>,
}
```

Where it has:

1. Address. Each module has to have address. TBD: Why, and how it is defined and used.
2. Several state attributes. In this case it is address of admin, and map of token id to owner address. Token id is u64 for simplicity.

TBD: What is context, what does it hold?

# Implementing `sov_modules_api::Module` trait
Before implementing `Module` trait, several things needs to be done:

1. Add new dependencies: serde, borsh, sov-state
2. Define 'native' feature flag to separate logic that is not needed in ZK context
3. Define Call and Query messages
4. Define Config

## Preparation

> 1. Define `native` feature in `Cargo.toml`:

Reviewer 2: Let's specify what `native` means: the logic runs outside of zkvm.

Reviewer 2: Ok it is explained below, closing the comment:)

# How to create new module for Sovereign Rollup

# Getting Started

Module is a separate crate that needs to implement `sov_modules_api::Module` trait and it's own change logic based on input messages.

Let's start with defining basic dependencies in `Cargo.toml`:

```toml
[dependencies]
anyhow = { anyhow = "1.0.62" }
sov-modules-api = { git = "https://github.com/Sovereign-Labs/sovereign.git", branch = "main", default-features = false }
sov-modules-macros = { git = "https://github.com/Sovereign-Labs/sovereign.git", branch = "main" }
```

## Module definition

> NFT module is defined as following:

Reviewer: This feels a little surprising - we haven't told readers _that_ we're building an NFT module yet, but we jump right into telling them how to do it. 

We also haven't explained where this code should live, or what the directory structure should look like. Maybe we should start by telling them that we'll follow the structure of the module template (with a link to the template that they can clone), and then we can add a note saying that this code replaces the `ExampleStruct` in `lib.rs`

```rust
use sov_modules_api::Context;
use sov_modules_macros::ModuleInfo;

#[derive(ModuleInfo, Clone)]
pub struct NonFungibleToken<C: Context> {
    #[address]
    pub address: C::Address,

    #[state]
    pub(crate) admin: sov_state::StateValue<C::Address>,

    #[state]
    pub(crate) owners: sov_state::StateMap<u64, C::Address>,
}
```

Where it has:

1. Address. Each module has to have an address. A module address is similar to a smart contract address in Ethereum. For example, a module can receive tokens. The address is calculated from a hash of the module prefix. This means that:
    - The module address is unique
    - No one knows the private key that produces this address

> 2. Several state attributes. In this case it is an address of admin, and map of token id to owner address. Token id is u64 for simplicity.

Reviewer: I think we need to give users some context about what `state` attributes are. It doesn't have to happen in this exact part of the tutorial, but we should at least let them know that we'll explain more about state values later. Broadly, things we want users to know about `state` values are that they are stored in the JMT, they can be read and written to, that the state values aren't part of a module (so the module can be cloned cheaply), that they're populated in a particular `WorkingSet`, etc.)

> 

Reviewer: It would be good to mention modules can also use `#[module]` to store a reference to another module, even though we don't use that feature in this example.

Author: Done

The Spec trait allows rollups to easily tailor themselves to different ZK VMs. 
By being generic over a Spec, a rollup can ensure that any potentially SNARK-unfriendly cryptography can be easily swapped out.

# Implementing `sov_modules_api::Module` trait
Before implementing `Module` trait, several things needs to be done:

1. Add new dependencies: serde, borsh, sov-state
2. Define 'native' feature flag to separate logic that is not needed in ZK context
3. Define Call and Query messages
4. Define Config

## Preparation

1. Define `native` feature in `Cargo.toml`:

```toml
[dependencies]
anyhow = "1.0.62"
borsh = { version = "0.10.3", features = ["bytes"] }
serde = { version = "1", features = ["derive"] }
serde_json = "1"

sov-modules-api = { git = "https://github.com/Sovereign-Labs/sovereign.git", branch = "main", default-features = false }
sov-modules-macros = { git = "https://github.com/Sovereign-Labs/sovereign.git", branch = "main" }
sov-state = { git = "https://github.com/Sovereign-Labs/sovereign.git", branch = "main", default-features = false }

[features]
default = ["native"]
serde = ["dep:serde", "dep:serde_json"]

> native = ["serde", "sov-state/native", "sov-modules-api/native"]

Reviewer 3: When reading I was a bit unclear how "defining" the native flag look like. 

After the "Define 'native' feature flag to separate logic that is not needed in ZK context" would it make sense to explain why in this instance the native flag definition includes "serde", and "sov-state/native" etc.?

Author: Addressed this

```

> It is needed for optimizing module for execution in ZK context, where none of RPC related logic is needed.

Reviewer 3: ```suggestion
The native flag is needed for optimizing module for execution in ZK context, where none of RPC related logic is needed.
```

2. Define Call and Query messages. Call messages are used to change state of the module, and Query messages are used to read state of the module.

```rust
use sov_modules_api::Context;

#[cfg_attr(feature = "native", derive(serde::Serialize), derive(serde::Deserialize))]
#[derive(borsh::BorshDeserialize, borsh::BorshSerialize, Debug, PartialEq, Clone)]
pub enum CallMessage<C: Context> {
    Mint {
        /// The id of new token. Caller is an owner
        id: u64
    },
    Transfer {
        /// The address to which the token will be transferred.
        to: C::Address,
        /// The token id to transfer
        id: u64,
    },
    Burn {
        id: u64,
    }
}

/// This enumeration responsible for querying the nft module.
#[derive(borsh::BorshDeserialize, borsh::BorshSerialize, Debug, PartialEq)]
pub enum QueryMessage {
    GetOwner { token_id: u64 },
}
```

> Query message is only used in native mode, because it does not change state in any way, since it is not used in state transition function.

Reviewer 3: For clarity, how about? 

```suggestion
The query message is exclusively used in native mode as it does not alter the state. This is because it is not utilized in the state transition function.
```

Reviewer 3: BTW, just curious, how do different modules read the state of other modules? Do they not use the query message?

Author: Module should expose public methods, using query messages adds overhead on serialization/deserialization.
